File: 00401. Binary Watch/solution.cpp
/**
 * Problem: 401. Binary Watch
 *
 * Difficulty: Easy
 *
 * Language: C++
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

class Solution {
public:
    vector<string> readBinaryWatch(int turnedOn) {
        // Handle edge cases
        if (turnedOn == 0)
            return {"0:00"};
        if (turnedOn > 8)
            return {};

        vector<string> validTimes;

        // Initialize bitmask with turnedOn least significant bits set to 1
        // Example: if turnedOn=3, mask = 111 (binary)
        int currentMask = (1 << turnedOn) - 1;

        // Generate all combinations of turnedOn bits among 10 positions
        while (currentMask < (1 << 10)) {
            // Check if current combination represents valid time
            string time = isValidTime(currentMask);
            if (!time.empty())
                validTimes.push_back(time);

            // Gosper's hack: find next combination with same number of set bits
            // Algorithm steps:
            // 1. Find the rightmost contiguous block of 1s
            // 2. Move the leftmost bit of that block one position left
            // 3. Shift the remaining bits to the far right
            int rightmostBit = currentMask & -currentMask;
            int nextMask = currentMask + rightmostBit;
            currentMask =
                (((nextMask ^ currentMask) >> 2) / rightmostBit) | nextMask;
        }

        return validTimes;
    }

    // Validates if a bitmask represents a valid time on binary watch
    // First 6 bits (0-5) represent minutes, next 4 bits (6-9) represent hours
    string isValidTime(int mask) {
        // Extract hour from bits 6-9
        int hour = mask >> 6;

        // Extract minute from bits 0-5
        int minute = mask & 63; // 63 = 111111 (6 bits)

        // Check if time is valid
        if (hour >= 12 || minute >= 60)
            return "";

        // Format time string with leading zero for minutes
        return to_string(hour) + ":" + (minute < 10 ? "0" : "") +
               to_string(minute);
    }
};

File: 00401. Binary Watch/solution.cs
/**
 * Problem: 401. Binary Watch
 *
 * Difficulty: Easy
 *
 * Language: C#
 *
 * Performance: Runtime 0 ms (Beats 100%)
 */

public class Solution
{
  private void GenerateValidHours(
    List<string> validTimes,
    int currentLedIndex,
    int ledsToUse,
    int currentHour,
    int currentMinute
  )
  {
    // Base case: used all LEDs
    if (ledsToUse == 0)
    {
      // Format minute with leading zero if needed
      string minuteString =
        currentMinute < 10 ? "0" + currentMinute.ToString() : currentMinute.ToString();

      // Add valid time to result list
      validTimes.Add(currentHour + ":" + minuteString);

      // Backtrack
      return;
    }

    // Skip current LED without using it
    if (currentLedIndex > ledsToUse - 1)
      GenerateValidHours(validTimes, currentLedIndex - 1, ledsToUse, currentHour, currentMinute);

    // Track new hour and minute values
    int newMinute = currentMinute;
    int newHour = currentHour;

    // Determine if current LED belongs to hour or minute section
    // LEDs 0-5 represent minutes, LEDs 6-9 represent hours
    if (currentLedIndex > 5)
      newHour += (int)Math.Pow(2, currentLedIndex - 6); // Add to hour
    else
      newMinute += (int)Math.Pow(2, currentLedIndex); // Add to minute

    // Use current LED only if resulting time is valid
    if (newHour < 12 && newMinute < 60)
      GenerateValidHours(validTimes, currentLedIndex - 1, ledsToUse - 1, newHour, newMinute);
  }

  public IList<string> ReadBinaryWatch(int turnedOn)
  {
    // If more than 9 LEDs are on, no valid time exists
    if (turnedOn > 9)
      return new List<string>();

    // List to store all valid times
    List<string> validTimes = new List<string>();

    // Generate all valid times using backtracking
    GenerateValidHours(validTimes, 9, turnedOn, 0, 0);

    // Return list of valid times
    return validTimes;
  }
}
